File: tests/leetcode/x/erlang/0006.erl
#!/usr/bin/env escript

convert(S, NumRows) when NumRows =< 1; NumRows >= length(S) -> S;
convert(S, NumRows) ->
    Cycle = 2 * NumRows - 2,
    lists:append([row_chars(S, NumRows, Row, Cycle) || Row <- lists:seq(0, NumRows - 1)]).

row_chars(S, NumRows, Row, Cycle) ->
    row_chars(S, NumRows, Row, Cycle, Row, []).
row_chars(S, NumRows, Row, Cycle, I, Acc) when I < length(S) ->
    Ch = lists:nth(I + 1, S),
    Diag = I + Cycle - 2 * Row,
    Acc1 = [Ch | Acc],
    Acc2 = case Row > 0 andalso Row < NumRows - 1 andalso Diag < length(S) of
        true -> [lists:nth(Diag + 1, S) | Acc1];
        false -> Acc1
    end,
    row_chars(S, NumRows, Row, Cycle, I + Cycle, Acc2);
row_chars(_, _, _, _, _, Acc) -> lists:reverse(Acc).

trim_cr(Line) -> lists:reverse(drop_cr(lists:reverse(Line))).
drop_cr([$\r | T]) -> T;
drop_cr(L) -> L.

main(_) ->
    {ok, Data} = file:read_file("/dev/stdin"),
    Lines0 = string:split(binary_to_list(Data), "\n", all),
    Lines = [trim_cr(L) || L <- Lines0],
    case Lines of
        [] -> ok;
        [First | Rest] ->
            T = list_to_integer(string:trim(First)),
            Outs = build(Rest, T, []),
            io:format("~s", [string:join(lists:reverse(Outs), "\n")])
    end.

build(_, 0, Acc) -> Acc;
build([S, Rows | Rest], T, Acc) ->
    build(Rest, T - 1, [convert(S, list_to_integer(string:trim(Rows))) | Acc]);
build(_, _, Acc) -> Acc.

File: tests/leetcode/x/elixir/0006.ex
defmodule Main do
  def convert(s, num_rows) do
    if num_rows <= 1 or num_rows >= byte_size(s) do
      s
    else
      cycle = 2 * num_rows - 2

      0..(num_rows - 1)
      |> Enum.map(fn row ->
        Enum.reduce(Stream.iterate(row, &(&1 + cycle)) |> Enum.take_while(&(&1 < byte_size(s))), "", fn i, acc ->
          acc = acc <> binary_part(s, i, 1)
          diag = i + cycle - 2 * row
          if row > 0 and row < num_rows - 1 and diag < byte_size(s) do
            acc <> binary_part(s, diag, 1)
          else
            acc
          end
        end)
      end)
      |> Enum.join("")
    end
  end

  def main do
    lines = IO.read(:stdio, :eof) |> String.split(~r/\r?\n/, trim: false)
    if lines != [] do
      t = hd(lines) |> String.trim() |> String.to_integer()
      {_, _, out} =
        Enum.reduce(1..t, {1, lines, []}, fn _, {idx, lines, out} ->
          s = Enum.at(lines, idx, "")
          num_rows = Enum.at(lines, idx + 1, "1") |> String.trim() |> String.to_integer()
          {idx + 2, lines, out ++ [convert(s, num_rows)]}
        end)
      IO.write(Enum.join(out, "\n"))
    end
  end
end

Main.main()
